tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc

```diff
@@ -1871,6 +1871,102 @@ class FuseSquaredDiffStage : public ArithmeticOptimizerStage {
   }
 };
 
+
+// Performs the conversions:
+// Add(Mul(x,y), z) => FusedMulAdd(Identity(x),y,z)
+// Add(Mul(x,y), Mul(z,t)) => FusedMulAdd(Identity(x),y,Identity(z),t)
+class FuseMulAddStage : public ArithmeticOptimizerStage {
+ public:
+  explicit FuseMulAddStage(const GraphOptimizerContext& ctx,
+                                const ArithmeticOptimizerContext& ctx_ext)
+      : ArithmeticOptimizerStage("FuseMulAddStage", ctx, ctx_ext) {}
+  ~FuseMulAddStage() override = default;
+
+  bool IsSupported(const NodeDef* node) const override {
+    if(!IsAdd(*node) && !IsSub(*node))
+      return false; 
+    // todo: handle rank>5
+    auto dtype = GetDataTypeFromAttr(*node, "T");
+    return (dtype==DT_HALF || dtype==DT_FLOAT || dtype==DT_DOUBLE);
+  }
+
+  Status TrySimplify(NodeDef* node, string* simplified_node_name) override {
+    NodeDef* b, *c;
+    TF_RETURN_IF_ERROR(GetInputNode(node->input(0), &b));
+    TF_RETURN_IF_ERROR(GetInputNode(node->input(1), &c));
+    auto dtype = GetDataTypeFromAttr(*node, "T");
+    bool add = IsAdd(*node);
+    bool can_absorb_b = (IsMul(*b) && !IsInPreserveSet(*b) &&
+        (NumNonControlOutputs(*b, *ctx().node_map) == 1));
+    bool can_absorb_c = (IsMul(*c) && !IsInPreserveSet(*c) &&
+        (NumNonControlOutputs(*c, *ctx().node_map) == 1));
+    if(GetDataTypeFromAttr(*b, "T") != dtype
+      ||GetDataTypeFromAttr(*c, "T") != dtype) {
+      VLOG(1)<<"Input dtype mismatch"; // apparently this can happen
+      return Status::OK();
+    }
+
+    if(can_absorb_b) {
+      NodeDef *b0, *b1;
+      TF_RETURN_IF_ERROR(GetInputNode(b->input(0), &b0));
+      TF_RETURN_IF_ERROR(GetInputNode(b->input(1), &b1));
+      auto dtype_b0 = GetDataTypeFromAttr(*b0, "T");
+      auto dtype_b1 = GetDataTypeFromAttr(*b1, "T");
+      if(dtype_b0!=dtype || dtype_b1!=dtype) {
```

Comment:
can this actually happen? 

given the registration of the `Mul` op here (https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/core/ops/math_ops.cc#L467 ), the output datatype and the two inputs datatypes are supposed to be the same, and there is already a check above to ensure the output datatype is a match

Reply:
I haven't looked at registrations, but I've seen that inputs of 'node' can be different types from its output, so checking the same is true for b and c seems prudent.